### PR TITLE
Serialize metadata as JSON

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
jenkins

File: src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala [hunk 1]
```diff
@@ -37,70 +40,63 @@ object VariantSampleMatrix {
...
-      Parser.parseType(schema)
-    }
+    val sampleInfoSchema = TStruct(("id", TString), ("annotation", saSignature))
+    val sampleInfo = getAndCastJSON[JArray]("sample metadata", "JArray")
+      .arr
+      .map {
+        case JObject(List(("id", JString(id)), ("annotation", jv: JValue))) =>
+          (id, JSONAnnotationImpex.importAnnotation(jv, saSignature, "sample-info"))
+        case other =>
+          fatal(s"invalid sample metadata: expected JObject with fields `id' and `annotation', but got `$other'")
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala]: This isn't sample metadata, it is sample annotations.

File: src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala [hunk 2]
```diff
@@ -37,70 +40,63 @@ object VariantSampleMatrix {
...
+      case jo: JObject => jo.obj.toMap
+      case _ => fatal("corrupt VDS: invalid metadata json.\n  Recreate VDS with current version of Hail.")
+    }
 
-        } catch {
-          case e: Exception =>
-            println(e)
-            fatal(s"Invalid VDS: ${e.getMessage}\n  Recreate with current version of Hail.")
-        }
+    def getAndCastJSON[T <: JValue](fname: String, expected: String)(implicit tct: ClassTag[T]): T =
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala]: Don't pass expected, use `tct.runtimeClass.getName` to match `other.getClass.getName`.

File: src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala [hunk 3]
```diff
@@ -37,70 +40,63 @@ object VariantSampleMatrix {
...
-    if (requireParquetSuccess && !hadoopExists(hConf, pqtSuccess))
-      fatal("corrupt VDS: no parquet success indicator, meaning a problem occurred during write.  Recreate VDS.")
+    val wasSplit = getAndCastJSON[JBool]("split", "JBool").value
+    val version = getAndCastJSON[JInt]("version", "JBool").num
 
-    if (!hadoopExists(hConf, vaSchema, saSchema, globalSchema))
-      fatal("corrupt VDS: one or more .schema files missing.  Recreate VDS.")
+    val saSignature = Parser.parseType(getAndCastJSON[JString]("sample annotation schema", "JString").s)
+    val vaSignature = Parser.parseType(getAndCastJSON[JString]("variant annotation schema", "JString").s)
+    val globalSignature = Parser.parseType(getAndCastJSON[JString]("global annotation schema", "JString").s)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala]: You're OK with spaces in the ids?

File: src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala [hunk 4]
```diff
@@ -37,70 +40,63 @@ object VariantSampleMatrix {
...
-      Parser.parseType(schema)
-    }
+    val globalAnnotation = JSONAnnotationImpex.importAnnotation(getAndCastJSON[JValue]("global annotation", "_"), globalSignature, "global")
 
-    val globalSignature = readFile(dirname + "/global.schema", hConf) { dis =>
-      val schema = Source.fromInputStream(dis)
-        .mkString
-      Parser.parseType(schema)
-    }
+    if (!hadoopExists(hConf, pqtSuccess) && requireParquetSuccess)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala]: Do this at the top with the other exists tests.

**Comment by cseed** [comment]
Delete `magicNumber`.  Bump the version number.  You should still check the version on import.

Back to you.

**Comment by tpoterba** [comment]
jenkins

**Comment by tpoterba** [comment]
jenkins

**Pull request merged**